### PR TITLE
[DebuggerV2] Sketch out TimelineComponent for Execution digests

### PR DESCRIPTION
**Pull request by caisq**
- Motivation for features / changes
  - Continue developing DebuggerV2 plugin.

- Technical description of changes
    - On loading, requests top-level execution events from the backend
      route /executions/digests
    - Once the number of executions are requested, the first page of
      ExecutionDigests are loaded from the same route and displayed in
      the newly-added TimelineComponent.
      - The size of the page is determined by how many ExecutionDigests
        are displayed by the TimelineComponent at a time (50 by default)
      - The request to the backend asks for more than 50 (1000 by default)
        to prevent the frontend from sending too many requests.
    - The ExecutionDigests are displayed with different CSS classes depending
      on whether they are individual ops such as `MatMul` or TensorFlow FuncGraphs
      such as `__inference_distributed_function_1235`
    - The user can use the left (<) and right (>) navigation buttons to
      navigate along the timeline. When necessary, new request to the
      /executions/digests route will be made so that the frontend can
      have the necessary data to display.
    - Adds all necessary actions, store states, selectors, reducers and
      effects to support the aforementioned frontend features.
  - Also in this CL:
    - Simplify the JSON-logic in debug_data_multiplexer.py by using the
      `ExecutionDigest.to_json()` method, which became available in tf-nightly today.

- Screenshots of UI changes
  1. 
image
  2. 
image

- Detailed steps to verify changes work correctly (as executed by you)
  - Unit tests are added (about 50% of the code in this PR are tests)
  - Verified using a few actual tfdbg2-format debug data dumps.

### CODE REVIEW

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/debugger_component.ts [hunk 1]
```diff
@@ -12,18 +12,41 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+
+  @Input()
+  activeRunId: string | null = null;
+
+  @Output()
+  onActiveRunIdChange = new EventEmitter<string>();
+
+  ngOnChanges(changes: SimpleChanges): void {
+    if (changes['activeRunId'] && changes['activeRunId'].currentValue) {
+      this.onActiveRunIdChange.emit(changes['activeRunId']
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/debugger_component.ts]: This feels very wrong. The data that comes from store is dispatched back to the flux. Why do you need to do this?

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts [hunk 2]
```diff
@@ -34,6 +43,66 @@ import {Tfdbg2HttpServerDataSource} from '../data_source/tfdbg2_data_source';
...
+): number[] {
+  if (pageSize <= 0 || !Number.isInteger(pageSize)) {
+    throw new Error(`Invalid pageSize: ${pageSize}`);
+  }
+  if (end > numItems) {
+    throw new Error(
+      `end index (${end}) exceeds total number of items (${numItems})`
+    );
+  }
+  if (end - begin > pageSize) {
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts]: this constraint feels weird to me. We are finding "missing pages" but `end - begin` cannot exceed `pageSize`?

Comment (stephanwlee) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts]: After reading the method more, I think it should read more like `getIncompletePages`?

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts [hunk 3]
```diff
@@ -34,6 +43,66 @@ import {Tfdbg2HttpServerDataSource} from '../data_source/tfdbg2_data_source';
...
+  }
+
+  // The constraint that `end - begin <= page` ensures that at most only two
+  // pages need to be requested.
+  const missingPages: number[] = [];
+  const pageIndex0 = Math.floor(begin / pageSize);
+  const anyDigestMissing0 =
+    !(pageIndex0 in pageLoadedSizes) ||
+    (pageLoadedSizes[pageIndex0] < pageSize &&
+      pageIndex0 * pageSize + pageLoadedSizes[pageIndex0] < numItems);
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts]: readability: consider factoring this conditional out and put more descriptive name around the logic.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts [hunk 4]
```diff
@@ -34,6 +43,66 @@ import {Tfdbg2HttpServerDataSource} from '../data_source/tfdbg2_data_source';
...
+    );
+  }
+  if (end - begin > pageSize) {
+    throw new Error('begin-end span exceeds page size, which is not allowed');
+  }
+
+  // The constraint that `end - begin <= page` ensures that at most only two
+  // pages need to be requested.
+  const missingPages: number[] = [];
+  const pageIndex0 = Math.floor(begin / pageSize);
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts]: readability: `pageIndex0` -> `beginPageIndex` or `pageIndexBegin`?

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts [hunk 5]
```diff
@@ -62,6 +131,77 @@ export class DebuggerEffects {
...
 
+  /** @export */
+  readonly loadNumExecutions$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(activeRunIdChanged),
+      withLatestFrom(this.store.select(getNumExecutionsLoaded)),
+      filter(([, loaded]) => loaded.state !== DataLoadState.LOADING),
+      tap(() => this.store.dispatch(numExecutionsRequested())),
+      mergeMap((props) => {
+        const runId = props[0].activeRunId as string; // TODO(cais): Guard against null?
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts]: `// TODO(cais): Guard against null?` Can we do this in the `filter` above? Also, what's the condition for this?

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/actions/debugger_actions.ts [hunk 6]
```diff
@@ -47,3 +50,42 @@ export const debuggerRunsRequestFailed = createAction(
...
+export const numExecutionsRequested = createAction(
+  '[Debugger] Number of Executions Requested'
+);
+
+export const numExecutionsLoaded = createAction(
+  '[Debugger] Number of Executions Loaded',
+  props<{numExecutions: number}>()
+);
+
+export const requestExecutionDigests = createAction(
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/actions/debugger_actions.ts]: Please fix: this action is fired from the UI and this is not describing user's action onto the view. It is describing how view understood user's interaction and converting it to the system's reaction to it.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/actions/debugger_actions.ts [hunk 7]
```diff
@@ -47,3 +50,42 @@ export const debuggerRunsRequestFailed = createAction(
...
+  '[Debugger] Request for ExecutionDigests',
+  props<{runId: string; begin: number; end: number; pageSize: number}>()
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/actions/debugger_actions.ts]: In effects, you are using `props.pageLoadedSizes` which SHOULD violate the typing here but it is not reporting as such. Can you check why? Also, an action should try hard not to contain any data that is accessible from the store directly (it should have user's inputs)

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts [hunk 8]
```diff
@@ -93,4 +98,398 @@ describe('Debugger reducers', () => {
...
+          start_time: 111,
+        },
+      },
+      runsLoaded: {
+        state: DataLoadState.LOADED,
+        lastLoadedTimeInMs: 222,
+      },
+      activeRunId: '__default_debugger_run__',
+    });
+    let nextState = reducers(state, actions.numExecutionsRequested());
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts]: I find this to be awkward. Do you need to fire another action beforehand in order to set up the environment in order to test the action? Is it possible to create the right debugger state instead?

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts [hunk 9]
```diff
@@ -93,4 +98,398 @@ describe('Debugger reducers', () => {
...
+        },
+      },
+      runsLoaded: {
+        state: DataLoadState.LOADED,
+        lastLoadedTimeInMs: 222,
+      },
+      activeRunId: '__default_debugger_run__',
+    });
+    const nextState = reducers(state, actions.numExecutionsRequested());
+    expect(nextState.executions.numExecutionsLoaded.state).toEqual(
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts]: Can you ever go from LOADED to LOADING? A spec would be nice :)

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts [hunk 10]
```diff
@@ -93,4 +98,398 @@ describe('Debugger reducers', () => {
...
+        },
+        executionDigestsLoaded: {
+          numExecutions: 1337,
+          pageLoadedSizes: {},
+          state: DataLoadState.NOT_LOADED,
+          lastLoadedTimeInMs: null,
+        },
+      }),
+    });
+    let nextState = reducers(state, actions.executionDigestsRequested());
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts]: let -> const.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts [hunk 11]
```diff
@@ -93,4 +98,398 @@ describe('Debugger reducers', () => {
...
+    ).toBeNull();
+    expect(nextState.executions.executionDigestsLoaded.numExecutions).toEqual(
+      1337
+    );
+    expect(nextState.executions.executionDigestsLoaded.pageLoadedSizes).toEqual(
+      {}
+    );
+  });
+
+  it('On executionDigestsLoaded: Empty --> complete 1st page', () => {
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts]: write expected behavior, not the setting of the test.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/views/timeline/timeline_component.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+  displayExecutionDigests: ExecutionDigestForDisplay[] = [];
+
+  @Output()
+  onNavigateLeft = new EventEmitter();
+
+  @Output()
+  onNavigateRight = new EventEmitter();
+
+  navigateLeft(): void {
+    this.onNavigateLeft.emit();
```
[the remark this answers is not in the document]
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/views/timeline/timeline_component.ts]: Done. Thanks.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects_test.ts [hunk 13]
```diff
@@ -21,13 +21,155 @@ import {ReplaySubject, of} from 'rxjs';
...
+  });
+
+  it('returns empty page indices for existing incomplete page', () => {
+    const begin = 80;
+    const end = 130;
+    const pageSize = 100;
+    const numItems = 1001;
+    const loadePageSizes = {
+      0: 100,
+      1: 30,
```
[the remark this answers is not in the document]
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects_test.ts]: Done.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts (hunk 9 above)
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts]: Yes. I added a spec in debugger_types.ts.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts (hunk 10 above)
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts]: Done. Did a pass to fix all oversights like this in this file.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts (hunk 11 above)
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts]: Done.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts (hunk 8 above)
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers_test.ts]: Yes. Done.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/views/timeline/timeline_container.ts [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+  '__backward_',
+  '__inference_',
+];
+
+/**
+ * Get a display version of the execution digest.
+ * @param executionDigest
+ * @param strLen
+ */
+export function getExecutionDigestForDisplay(
```
[the remark this answers is not in the document]
Comment (stephanwlee) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/views/timeline/timeline_container.ts]: Are you missing a commit here? I do not see the change in code.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts [hunk 15]
```diff
@@ -62,9 +139,152 @@ export class DebuggerEffects {
...
+          return requestExecutionDigests({
+            runId: runId!,
+            begin,
+            end,
+            pageSize,
+          });
+        }
+      )
+    )
+  );
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts]: I would like to propose major change to this block of code. These two effects are interacting with each other using whole flux cycles. Instead, how about this?

```ts

   private readonly digestRequired$ = this.actions$.pipe(
    ofType(executionScrollLeft, executionScrollRight),
    withLatestFrom(
      this.store.select(getActiveRunId),
      this.store.select(getExecutionScrollBeginIndex),
      this.store.select(getNumExecutions),
      this.store.select(getDisplayCount),
      this.store.select(getExecutionPageSize)
    ),
    filter(data => {
      const runId = data[1];
      return runId !== null;
    }),
    map(
      ([_, runId, scrollBeginIndex, numExecutions, displayCount, pageSize]) => {
        const begin = scrollBeginIndex;
        const end = Math.min(numExecutions, begin + displayCount);
        return {
          runId: runId!,
          begin,
          end,
          pageSize
        };
      }
    )
  );

  readonly loadExecutionDigests$ = createEffect(() =>
    digestRequired$.pipe(
      withLatestFrom(this.store.select(getExecutionDigestsLoaded)),
      filter(([_, loaded]) => loaded.state !== DataLoadState.LOADING),
      map(([props, loaded]) => {
        return {
          props,
          loaded,
          missingPages: getMissingPages(
            props.begin,
            props.end,
            props.pageSize,
            loaded.numExecutions,
            loaded.pageLoadedSizes
          )
        };
      }),
      filter(({ missingPages }) => missingPages.length > 0),
      tap(() => this.store.dispatch(executionDigestsRequested())),
      withLatestFrom(this.store.select(getExecutionDigestsLoaded)),
      mergeMap(([{ props, missingPages }, loaded]) => {
        const { runId, pageSize } = props;
        const actualBegin = missingPages[0] * pageSize;
        const actualEnd = Math.min(
          loaded.numExecutions,
          (missingPages[missingPages.length - 1] + 1) * pageSize
        );
        return this.dataSource
          .fetchExecutionDigests(runId, actualBegin, actualEnd)
          .pipe(
            map(
              digests => {
                return executionDigestsLoaded(digests);
              }
              // TODO(cais): Add catchError() to pipe.
            )
          );
      })
    )
  );

  // ALTERNATIVE
  readonly loadExecutionDigests2$ = createEffect(() =>
    this.actions$.pipe(
      ofType(executionScrollLeft, executionScrollRight),
      withLatestFrom(
        this.store.select(getExecutionDigestsLoaded),
        this.store.select(getActiveRunId),
        this.store.select(getExecutionScrollBeginIndex),
        this.store.select(getNumExecutions),
        this.store.select(getDisplayCount),
        this.store.select(getExecutionPageSize)
      ),
      filter(
        ([_, loaded, runId]) =>
          loaded.state !== DataLoadState.LOADING && runId !== null
      ),
      map(args => {
        const [
          ,
          loaded,
          runId,
          scrollBeginIndex,
          numExecutions,
          displayCount,
          pageSize
        ] = args;
        const begin = scrollBeginIndex;
        const end = Math.min(numExecutions, begin + displayCount);
        return {
          runId,
          scrollBeginIndex,
          numExecutions,
          displayCount,
          pageSize,
          loaded,
          missingPages: getMissingPages(
            begin,
            end,
            pageSize,
            loaded.numExecutions,
            loaded.pageLoadedSizes
          )
        };
      }),
      filter(({ missingPages }) => missingPages.length > 0),
      tap(() => this.store.dispatch(executionDigestsRequested())),
      mergeMap(({loaded, missingPages, runId, pageSize}) => {
        const actualBegin = missingPages[0] * pageSize;
        const actualEnd = Math.min(
          loaded.numExecutions,
          (missingPages[missingPages.length - 1] + 1) * pageSize
        );

        return this.dataSource
          .fetchExecutionDigests(runId, actualBegin, actualEnd)
          .pipe(
            map(
              digests => {
                return executionDigestsLoaded(digests);
              }
              // TODO(cais): Add catchError() to pipe.
            )
          );
      })
    )
  );
```

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts [hunk 16]
```diff
@@ -62,9 +139,152 @@ export class DebuggerEffects {
...
+            props.end,
+            props.pageSize,
+            loaded.numExecutions,
+            loaded.pageLoadedSizes
+          ),
+        };
+      }),
+      filter(({missingPages}) => missingPages.length > 0),
+      tap(() => this.store.dispatch(executionDigestsRequested())),
+      withLatestFrom(this.store.select(getExecutionDigestsLoaded)),
```
Comment (stephanwlee) [inline tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts]: Please just consider using `loaded` from L205 instead. That would make argument to mergeMap simpler, too.

**Comment by caisq** [comment]
cc @baileydesign

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts [hunk 17]
```diff
@@ -58,13 +83,168 @@ const reducer = createReducer(
...
+          executionDigestsLoaded: {
+            ...state.executions.executionDigestsLoaded,
+            numExecutions: digests.num_digests,
+            state: DataLoadState.LOADED,
+            lastLoadedTimeInMs: Date.now(),
+          },
+          executionDigests: Object.assign(
+            {},
+            state.executions.executionDigests
+          ),
```
[the remark this answers is not in the document]
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts]: Done.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/debugger_container_test.ts [hunk 18]
```diff
@@ -128,4 +140,152 @@ describe('Debugger Container test', () => {
...
+
+      const navigationPositionInfoElement = fixture.debugElement.query(
+        By.css('.navigation-position-info')
+      );
+      expect(navigationPositionInfoElement.nativeElement.innerText).toBe(
+        'Execution: 977 ~ 1076 of 1500'
+      );
+    });
+
+    it('left-button click dispatches executionScrollLeft action', async () => {
```
[the remark this answers is not in the document]
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/debugger_container_test.ts]: Done. Also did a pass to remove unnecessary `async` in this test file.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts (hunk 16 above)
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts]: Done. Thanks for catching this.

File: tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts (hunk 15 above)
Comment (caisq) [reply tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts]: Done. Adopted your first suggestion. It simplifies the code and gets ride of an action type. Thanks.

**Pull request merged**